### PR TITLE
net: lib: download client: add IPv6 and HTTPS

### PR DESCRIPTION
**Pull request by lemrey**
- Added IPv6 support
When enabled prefers IPv6 connections and fallback to IPv4.
- Added HTTPS support
The application must provision the TLS credentials and provide the security tag to the library.
- Handle errors while sending the HTTP request in the thread
- Updated documentation
- Cleaned up logs
- Updated changelog

### CODE REVIEW

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 1]
```diff
@@ -26,6 +26,29 @@ BUILD_ASSERT_MSG(CONFIG_NRF_DOWNLOAD_MAX_FRAGMENT_SIZE <=
...
+	struct timeval timeo = {
+		.tv_sec = CONFIG_NRF_DOWNLOAD_CLIENT_SOCK_TIMEOUT_MS,
+	};
+
+	LOG_INF("Configuring socket timeout (%ld ms)", timeo.tv_sec);
+
+	err = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeo, sizeof(timeo));
+	if (err) {
+		LOG_WRN("Failed to set socket timeout, errno %d", errno);
+		return -1;
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/http_download_client.c]: nit: You could return the actual error code here (`return -errno;`)

File: include/download_client.h [hunk 2]
```diff
@@ -33,7 +33,9 @@ extern "C" {
 enum download_client_evt_id {
 	/** Event contains a fragment. */
 	DOWNLOAD_CLIENT_EVT_FRAGMENT,
-	/** Download complete. */
+	/** Download complete.
+	 *  The socket is closed automatically.
```
Comment (rlubos) [inline include/download_client.h]: Is it? I haven't seen any code added that would do that.

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 3]
```diff
@@ -49,71 +56,126 @@ static int socket_timeout_set(int fd)
...
+	int verify;
+	sec_tag_t sec_tag_list[] = { sec_tag };
+
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	verify = OPTIONAL;
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/http_download_client.c]: Why optional? I think that in a proper application we should always verify peer certificate.

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 4]
```diff
@@ -49,71 +56,126 @@ static int socket_timeout_set(int fd)
...
+		LOG_ERR("Failed to setup peer verification, errno %d", errno);
+		return -1;
+	}
+
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+			 sizeof(sec_tag_t) * ARRAY_SIZE(sec_tag_list));
+	if (err) {
+		LOG_ERR("Failed to setup socket security tag, errno %d", errno);
+		return -1;
+	}
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/http_download_client.c]: An empty line would look nice here.

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 5]
```diff
@@ -49,71 +56,126 @@ static int socket_timeout_set(int fd)
...
 		}
 	}
 
-	freeaddrinfo(addrinf);
+	/* Not connected */
+	err = -1;
 
-	if (rc < 0) {
+	for (addr = info; addr != NULL; addr = addr->ai_next) {
+
```
Comment (rlubos) [inline subsys/net/lib/download_client/src/http_download_client.c]: No newline needed at the beginning of the block.

File: subsys/net/lib/download_client/Kconfig [hunk 6]
```diff
@@ -11,13 +11,15 @@ if NRF_DOWNLOAD_CLIENT
 
 config NRF_DOWNLOAD_MAX_FRAGMENT_SIZE
 	int "Fragment size"
-	default 2048
+	range 512 2048 if BSD_LIBRARY && NRF_DOWNLOAD_CLIENT_TLS
+	default 2048 if BSD_LIBRARY && NRF_DOWNLOAD_CLIENT_TLS
```
Comment (rlubos) [inline subsys/net/lib/download_client/Kconfig]: Should we add a word of explanation on why do we have this limitation?

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 1 above)
Comment (ioannisg) [reply subsys/net/lib/download_client/src/http_download_client.c]: Yeap, I agree

File: subsys/net/lib/download_client/Kconfig [hunk 7]
```diff
@@ -26,6 +26,14 @@ config NRF_DOWNLOAD_CLIENT_STACK_SIZE
 	int "Thread stack size"
 	default 2048
 
+config NRF_DOWNLOAD_CLIENT_SOCK_TIMEOUT_MS
+	int "Receive timeout, in millisec"
```
Comment (ioannisg) [inline subsys/net/lib/download_client/Kconfig]: can you write the whole word here, i.e. "milliseconds"

File: include/download_client.rst [hunk 8]
```diff
@@ -3,50 +3,53 @@
 Download client
 ###############
 
-The download client library provides functions to download resources from a remote server.
-The resource could, for example, be a firmware image.
-Currently, the only supported protocol for the download is HTTP.
+The download client library can be used to download files from a HTTP or
```
Comment (ioannisg) [inline include/download_client.rst]: a --> an

File: subsys/net/lib/download_client/Kconfig (hunk 6 above)
Comment (ioannisg) [reply subsys/net/lib/download_client/Kconfig]: Yes, nice. This should go in the help text, imho.

File: subsys/net/lib/download_client/Kconfig [hunk 9]
```diff
@@ -11,13 +11,15 @@ if NRF_DOWNLOAD_CLIENT
 
 config NRF_DOWNLOAD_MAX_FRAGMENT_SIZE
 	int "Fragment size"
-	default 2048
+	range 512 2048 if BSD_LIBRARY && NRF_DOWNLOAD_CLIENT_TLS
+	default 2048 if BSD_LIBRARY && NRF_DOWNLOAD_CLIENT_TLS
+	default 4096
```
Comment (ioannisg) [inline subsys/net/lib/download_client/Kconfig]: Also, this is an increase, compared to before. We need to have an explanation for that.

File: subsys/net/lib/download_client/Kconfig [hunk 10]
```diff
@@ -34,9 +36,25 @@ config NRF_DOWNLOAD_CLIENT_SOCK_TIMEOUT_MS
 	  Socket timeout for recv() calls, in milliseconds.
 	  Set to -1 disable.
 
+config NRF_DOWNLOAD_CLIENT_IPV6
+	bool "Prefer IPv6 connections"
+	help
+	  Favor IPv6 protocol but fallback to
```
Comment (ioannisg) [inline subsys/net/lib/download_client/Kconfig]: Prefer

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 11]
```diff
@@ -24,7 +25,13 @@ LOG_MODULE_REGISTER(download_client, CONFIG_NRF_DOWNLOAD_CLIENT_LOG_LEVEL);
 
 BUILD_ASSERT_MSG(CONFIG_NRF_DOWNLOAD_MAX_FRAGMENT_SIZE <=
 		 CONFIG_NRF_DOWNLOAD_MAX_RESPONSE_SIZE,
-	"The response buffer must accommodate for a full fragment");
+		 "The response buffer must accommodate for a full fragment");
+
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE)
+BUILD_ASSERT_MSG(IS_ENABLED(CONFIG_NRF_DOWNLOAD_CLIENT_LOG_HEADERS) ?
```
Comment (ioannisg) [inline subsys/net/lib/download_client/src/http_download_client.c]: Why is this build-assert message inside CONFIG_LOG?

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 12]
```diff
@@ -49,71 +56,126 @@ static int socket_timeout_set(int fd)
...
+	struct addrinfo *addr;
+	struct addrinfo *info;
 
 	if (host == NULL) {
 		return -1;
 	}
 
-	struct addrinfo *addrinf = NULL;
+	/* Set up port and protocol */
+	if (sec_tag < 0) {
```
Comment (ioannisg) [inline subsys/net/lib/download_client/src/http_download_client.c]: Why don't we check explicitly with -1?

File: subsys/net/lib/download_client/Kconfig (hunk 9 above)
Comment (lemrey) [reply subsys/net/lib/download_client/Kconfig]: Would you like a separate commit for that or can I add some text in `help` to explain why it is better to have a larger value when not using TLS?

Comment (rlubos) [reply subsys/net/lib/download_client/Kconfig]: I think it's fine to do it in the same commit that the change was introduced.

File: include/download_client.h (hunk 2 above)
Comment (lemrey) [reply include/download_client.h]: You are right, I got confused; the socket is never closed automatically, it is left to the application.
Fixed.

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 3 above)
Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: Resolved offline

Comment (ioannisg) [reply subsys/net/lib/download_client/src/http_download_client.c]: ok

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 5 above)
Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: I had added it for readability, removed now :)

File: subsys/net/lib/download_client/Kconfig (hunk 6 above)
Comment (lemrey) [reply subsys/net/lib/download_client/Kconfig]: Added an explicit sentence in `help`.

File: subsys/net/lib/download_client/Kconfig (hunk 7 above)
Comment (lemrey) [reply subsys/net/lib/download_client/Kconfig]: Done

File: include/download_client.rst (hunk 8 above)
Comment (lemrey) [reply include/download_client.rst]: Corrected

File: subsys/net/lib/download_client/Kconfig (hunk 9 above)
Comment (lemrey) [reply subsys/net/lib/download_client/Kconfig]: Added a sentence in `help`.

File: subsys/net/lib/download_client/Kconfig (hunk 10 above)
Comment (lemrey) [reply subsys/net/lib/download_client/Kconfig]: Changed

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 12 above)
Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: No reason in particular, changed to check against `-1`.

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 4 above)
Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: Done

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 1 above)
Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: Done!

File: subsys/net/lib/download_client/src/http_download_client.c (hunk 12 above)
Comment (ioannisg) [reply subsys/net/lib/download_client/src/http_download_client.c]: alright

File: include/download_client.rst [hunk 13]
```diff
@@ -3,50 +3,53 @@
...
-* The address family is IPv4.
-* TCP transport is used to communicate with the server.
 * The application protocol to communicate with the server is HTTP 1.1.
 * IETF RFC 7233 is supported by the HTTP Server.
-* :option:`CONFIG_NRF_DOWNLOAD_MAX_RESPONSE_SIZE` is configured so that it can contain the entire HTTP response.
+* :option:`CONFIG_NRF_DOWNLOAD_MAX_RESPONSE_SIZE` is configured so that it
+  can contain the entire HTTP response.
 
-HTTPS is not supported.
+HTTPS
```
Comment (sigvartmh) [inline include/download_client.rst]: Maybe add a note that only TLS 1.2 is supported?

File: subsys/net/lib/download_client/Kconfig (hunk 9 above)
Comment (ioannisg) [reply subsys/net/lib/download_client/Kconfig]: thanks

File: include/download_client.rst (hunk 13 above)
Comment (ioannisg) [reply include/download_client.rst]: @lemrey ? @rlubos FYI

Comment (rlubos) [reply include/download_client.rst]: Could be a good point, but I'd rather avoid another CI round, unless absolutely necessary. Docs will be a subject of change after we have RC1 tagged anyway.

**Comment by sigvartmh** [comment]
Another thing that should be changed if some changes are going in is `	LOG_INF("Downloading: %s [%u]", client->file, client->progress);
` 
It should be `LOG_INF("Downloading: %s [%u]", log_strdup(client->file), client->progress);`

File: subsys/net/lib/download_client/src/http_download_client.c [hunk 14]
```diff
@@ -428,10 +545,6 @@ int download_client_start(struct download_client *client, char *file,
 
 	LOG_INF("Downloading: %s [%u]", client->file, client->progress);
```
Comment (sigvartmh) [inline subsys/net/lib/download_client/src/http_download_client.c]:
```suggestion
	LOG_INF("Downloading: %s [%u]", log_strdup(client->file), client->progress);

Comment (sigvartmh) [reply subsys/net/lib/download_client/src/http_download_client.c]: Another thing that should be changed if some changes are going in is LOG_INF("Downloading: %s [%u]", client->file, client->progress);

Comment (ioannisg) [reply subsys/net/lib/download_client/src/http_download_client.c]: @lemrey ?

Comment (lemrey) [reply subsys/net/lib/download_client/src/http_download_client.c]: Done, thanks.

File: include/download_client.rst (hunk 13 above)
Comment (lemrey) [reply include/download_client.rst]: Added, thanks.

**Comment by lemrey** [comment]
Added a commit to address @ioannisg latest comments.

File: include/download_client.rst (hunk 13 above)
Comment (ioannisg) [reply include/download_client.rst]: Well, we do run CI now, so we could fix this too if needed

Comment (ioannisg) [reply include/download_client.rst]: Ah, ok, I saw it was added.

**Comment by sigvartmh** [comment]
looked good to me. Sry for late reply was out of office 👍

**Pull request merged**